text/3724-theupdateframework.md

```diff
@@ -0,0 +1,341 @@
+
+## TAP-16 Implementation
+
+We're proposing to use [TAP-16](https://github.com/theupdateframework/taps/blob/master/tap16.md) to provide efficient update checking and download sizes. TAP-16 uses Merkle trees rather than full lists for the download of a snapshot of the inventory of a repository (`snapshot.json`). We want to ensure that, as crates.io grows, the total size clients have to download when checking for updates remains small.
+
+### Shared local folder changes
+
+Creation of a new `~/.cargo/tuf` directory. (If Cargo stores its registry information in another directory, the `tuf` directory should be stored alongside the `registry` directory.) This directory shall be used for all TUF operations by project tools (both Rustup and Cargo). The cargo folder was chosen as the main location of residence for these files given that although Rustup will be performing the initialization of these folders, there is already a precedent set for shared files living within the cargo folder.
+
+- `~/.cargo/tuf` The top-level directory of local copies of TUF repositories
+- `~/.cargo/tuf/root` a copy of the `tuf-root` repository locally synchronized
+- `~/.cargo/tuf/crates` a copy of the `tuf-crates` repository locally synchronized
+
+### Controlling enablement of TUF
+
+This RFC does not specify how to handle non-`crates.io` repositories. Cargo can choose to enable TUF for third-party repositories in the future, or may default to only using TUF for crates.io unless otherwise configured. Cargo might choose to use an environment variable (e.g. `CARGO_TUF_DISABLE`) to disable all usage of TUF on third-party repositories.
+
+## Cargo changes
+- Addition of `cargo-tuf-lib` as a dependency
+- `cargo-tuf-lib::sync` attempted prior to an index update
+- `cargo-tuf-lib::verify_snapshot` called on an index update on the entire index
+
+## Rustup Changes
+
+- Addition of `cargo-tuf-lib` as a dependency
+- `cargo-tuf-lib::sync` attempted on execution
+- `cargo-tuf-lib::verify_snapshot` called on an index update on the entire index
+- `cargo-tuf-lib::verify_file` called on the download of any release files (stable and nightly), or self-updates
+
+## Crates.io changes
+
+- Prior to updating the index, crates.io shall perform the online signing of the index entry to update the targets and sign the index entry, saving this as a like-pathed artifact in the TUF repository.
+
+## Infra Changes
+
+The infrastructure team.
+- Creation of `rust-lang/tuf-root` and `rust-lang/tuf-crates` repositories on GitHub
+- Initiation of the root signing ceremony via tuf-on-ci on each repository
+- Facilitate the initial and subsequent signing events
+- Determine how to mirror and distribute both repositories via CDN. We recommend that synchronization of the repository for the `rust-tuf-lib` should be done via HTTPS downloads from the CDN to prevent the end-user downloads being reliant on `git` or `GitHub`
```

Comment:
`HTTP` there is no need for the `S`. The entire point of TUF is that you don't need to trust any of the computers you received the file from in order to verify that they are authentic and up-to-date.

Reply:
> `HTTP` there is no need for the `S`.

Does cargo have any privacy goals?

Reply:
I think I would still like to serve everything under rust-lang.org and crates.io as HTTPS.

Reply:
While it's technically possible to build a scheme using TUF that provides security (but not privacy) over HTTP, for this proposal we do want both, so we're going to stick to HTTPS.